### PR TITLE
feat: convert project to typescript, add types

### PR DESCRIPTION
**Pull request by wonderfulfrog**
## Description

This PR updates the project to use TypeScript, and include type definitions when importing the module in other projects.

The node version was also bumped to `20.16.0`, which is the current LTS version.

I did my best to break up each change into smaller commits to follow the flow of my changes. I borrowed a few files from `wethegit/react-modal` to make things easier (e.g. `utils/classnames.ts`).

## Notes

There is a warning about different versions of TypeScript when building the project:

```sh
$ npm run build
// Output truncated
Analysis will use the bundled TypeScript version 5.4.2
*** The target project appears to use TypeScript 5.5.4 which is newer than the bundled compiler engine; consider upgrading API Extractor.
```

There aren't any build errors as a result of this warning. [There is a PR open](https://github.com/microsoft/rushstack/pull/4815) that updates TypeScript to 5.5 for [API Extractor](https://github.com/microsoft/rushstack/tree/main/apps/api-extractor). I believe once that PR is merged the warning will disappear.

### CODE REVIEW

File: src/lib/marquee.tsx [hunk 1]
```diff
@@ -1,23 +1,40 @@
...
- * @param {number=50} props.speed - The speed of the marquee in pixels per second.
- * @param {number=20} props.reducedMotionSpeed - The speed of the marquee in pixels per second when prefersReducedMotion is true.
- * @param {boolean=false} props.prefersReducedMotion - If true, marquee will use reducedMotionSpeed instead of speed.
- * @param {boolean=true} props.playing - If false, the marquee will not animate.
- * @param {string} props.className - Additional classnames to add to the marquee.
- * @param {React.Component} props.children - The content to be rendered inside the marquee. Will be wrapped in a div with the class of `marquee__slide`.
- *
- * @returns {React.Component}
- */
+export interface MarqueeProps extends React.Component {
```
Comment (marlonmarcello) [inline src/lib/marquee.tsx:7]: This is bit more forgiving as it allows users of the component to pass any valid HTML attribute that a `div` can have.

Important as you can then pass `data-` and `aria-` attributes as needed.

```suggestion
export interface MarqueeProps extends React.ComponentPropsWithoutRef<"div"> {
```

File: src/lib/marquee.tsx [hunk 2]
```diff
@@ -1,23 +1,40 @@
...
+  /**
+   * The content to be rendered inside the marquee. Will be wrapped in a div with the class of `marquee__slide`.
+   */
+  children: ReactElement
+  /**
+   * Additional classnames to add to the marquee.
+   */
+  className: string
```
Comment (marlonmarcello) [inline src/lib/marquee.tsx:15]: If you accept the suggestion above these become redundant.

```suggestion
```

**Review by marlonmarcello — CHANGES_REQUESTED** [review]
Great work @wonderfulfrog !

I'd advice to test this locally in an external project by using `npm link`.

Just to make sure the build and the types output are as expected.

File: vite.config.ts
Comment (marlonmarcello) [inline vite.config.ts:1]: @wonderfulfrog while you are at it, can you please update the rollupOption to also externalize the react runtime like we have here on rect-modal please?
https://github.com/wethegit/react-modal/blob/main/vite.config.ts#L19

**Review by marlonmarcello — CHANGES_REQUESTED** [review]

File: vite.config.ts
Comment (wonderfulfrog) [reply vite.config.ts:1]: @marlonmarcello Did you also want me to include `"react-dom": "ReactDOM",` as part of the `globals` option?

**Review by andrewrubin — COMMENTED** [review]
Looks great @wonderfulfrog, thanks for taking the time to do this 👏

File: vite.config.ts
Comment (marlonmarcello) [reply vite.config.ts:1]: Yes please 🙏

Comment (wonderfulfrog) [reply vite.config.ts:1]: I added the "react/jsx-runtime" to the external array, like in `react-modal`. It created a warning though, until I added it was part of the `globals` property. Let me know if this is okay or if we should change it.

Comment (marlonmarcello) [reply vite.config.ts:1]: That's correct, it's exactly how the react-modal one

**Comment by wonderfulfrog** [comment]
Using `npm link`, I can confirm the latest revision of the build works with a fresh Next.js project.

A fresh corgi project is having issues. It might be a corgi-specific problem, I'm not sure. Here is an error log I'm getting:
image

```
TypeError: Cannot read properties of null (reading 'useState')
```

image
```
Error: There was an error while hydrating. Because the error happened outside of a Suspense boundary, the entire root will switch to client rendering.
```

**Comment by marlonmarcello** [comment]
The `corgi` issue is really weird and I have a feeling it might have to do with `npm link`

**Review by marlonmarcello — APPROVED** [review]

**Review by marlonmarcello — APPROVED** [review]

**Pull request merged**